lib/teracy-dev-core/providers/manager.rb

```diff
@@ -0,0 +1,40 @@
+require 'teracy-dev/logging'
+
+require_relative 'virtualbox'
+require_relative 'vmware'
+
+module TeracyDevCore
+  module Providers
+    class Manager
+      @@instance = nil
+
+      attr_accessor :provider
+
+      def initialize
+        if !!@@instance
+          raise "TeracyDevCore::Providers::Manager can only be initialized once"
+        end
+
+        @@instance = self
+        @logger = TeracyDev::Logging.logger_for(self.class.name)
+      end
+
+      def register(provider_type)
+        case provider_type
+        when 'virtualbox'
```

Comment:
we never know beforehand the type of providers, this is considered hard-code and not understanding the flow.